### PR TITLE
Relay Extract: Silo/Overlink

### PR DESCRIPTION
**Pull request by JustinPrivitera**

### CODE REVIEW

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp [hunk 1]
```diff
@@ -730,6 +755,112 @@ RelayIOSave::execute()
...
+        if (params().has_path("topologies"))
+        {
+            const conduit::Node &tlist = params()["topologies"];
+            const int num_topos = tlist.number_of_children();
+            if (num_topos != 1)
+            {
+                ASCENT_ERROR("relay_io_save Overlink save requires a single topology; " <<
+                             num_topos << " topologies were requested.");
+            }
+        }
```
Comment (JustinPrivitera) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp:780]: @cyrush Can I assume params are the same across all ranks?

Comment (cyrush) [reply src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp:780]: yes, pre condition is that all ranks provide the same set of actions, so you don't need to handle that case.

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp [hunk 2]
```diff
@@ -730,6 +755,112 @@ RelayIOSave::execute()
...
+                            unique_topos[topo_name];
+                        }
+                    }
+                }
+                const int num_topos = unique_topos.number_of_children();
+                if (num_topos != 1)
+                {
+                    error = 1;
+                    error_oss << "relay_io_save Overlink save requires a single topology; "
+                                 "there are " << num_topos << " topologies in the input mesh.";
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp]: for the error case, lets show them all the unique topology names as part of the error message, and also tell them the param name that they can use to select the topology.

File: src/tests/ascent/t_ascent_mpi_relay_extract.cpp [hunk 3]
```diff
@@ -966,6 +966,301 @@ TEST(ascent_relay, test_relay_mpi_sparse_topos_2)
...
+    else if(par_rank == 1)
+    {
+        data.remove(0);
+        data.remove(0);
+        data.remove(0);
+        data.remove(0);
+    }
+    else
+    {
+        // cyrus was wrong about 2 mpi ranks.
```
Comment (cyrush) [inline src/tests/ascent/t_ascent_mpi_relay_extract.cpp:1017]: I still love this comment

File: src/tests/ascent/t_ascent_mpi_relay_extract.cpp [hunk 4]
```diff
@@ -966,6 +966,301 @@ TEST(ascent_relay, test_relay_mpi_sparse_topos_2)
...
+        for(int i=0;i<nfiles_to_check;i++)
+        {
+            std::string fprefix = "file_";
+            if(nfiles_to_check == 7)
+            {
+                // in the n domains == n files case, the file prefix is
+                // domain_
+                fprefix = "domain_";
+            }
+            snprintf(fmt_buff, sizeof(fmt_buff), "%06d",i);
```
Comment (cyrush) [inline src/tests/ascent/t_ascent_mpi_relay_extract.cpp]: Let's use `conduit_fmt::format` instgead of snprintf

File: src/tests/ascent/t_ascent_mpi_relay_extract.cpp [hunk 5]
```diff
@@ -966,6 +966,301 @@ TEST(ascent_relay, test_relay_mpi_sparse_topos_2)
...
+        for(int i=0;i<nfiles_to_check;i++)
+        {
+            std::string fprefix = "domfile";
+            if(nfiles_to_check == 7)
+            {
+                // in the n domains == n files case, the file prefix is
+                // domain_
+                fprefix = "domain";
+            }
+            snprintf(fmt_buff, sizeof(fmt_buff), "%d",i);
```
Comment (cyrush) [inline src/tests/ascent/t_ascent_mpi_relay_extract.cpp]: `conduit_fmt::format`

File: src/tests/ascent/t_ascent_relay.cpp [hunk 6]
```diff
@@ -1117,6 +1117,203 @@ TEST(ascent_relay, test_relay_lor_extract)
...
+        for(int i=0;i<nfiles_to_check;i++)
+        {
+            std::string fprefix = "file_";
+            if(nfiles_to_check == 7)
+            {
+                // in the n domains == n files case, the file prefix is
+                // domain_
+                fprefix = "domain_";
+            }
+            snprintf(fmt_buff, sizeof(fmt_buff), "%06d",i);
```
Comment (cyrush) [inline src/tests/ascent/t_ascent_relay.cpp]: `conduit_format::fmt`

File: src/tests/ascent/t_ascent_relay.cpp [hunk 7]
```diff
@@ -1117,6 +1117,203 @@ TEST(ascent_relay, test_relay_lor_extract)
...
+        for(int i=0;i<nfiles_to_check;i++)
+        {
+            std::string fprefix = "domfile";
+            if(nfiles_to_check == 7)
+            {
+                // in the n domains == n files case, the file prefix is
+                // domain_
+                fprefix = "domain";
+            }
+            snprintf(fmt_buff, sizeof(fmt_buff), "%d",i);
```
Comment (cyrush) [inline src/tests/ascent/t_ascent_relay.cpp]: `conduit_fmt`

**Review by cyrush — CHANGES_REQUESTED** [review]
Looks good a few small suggestions!

File: src/tests/ascent/t_ascent_mpi_relay_extract.cpp (hunk 3 above)
Comment (JustinPrivitera) [reply src/tests/ascent/t_ascent_mpi_relay_extract.cpp:1017]: it lives on, across projects, space, and time

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp (hunk 2 above)
Comment (JustinPrivitera) [reply src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp]: How does this look?
```
[Error] Ascent::execute 
file: /usr/workspace/justin/ascent_builds/develop_08_22_24/ascent/src/libs/ascent/runtimes/flow_filters/ascent_runtime_relay_filters.cpp
line: 872
message: 
relay_io_save Overlink save requires a single topology; there are 3 topologies in the input mesh. The current topologies are 
 - topo
 - topo2
 - topo3
You can select which topologies to save using the "topologies" parameter like in the following example:
extracts["<extract_name>/params/topologies"].append() = "<topo_name>"
```

**Review by cyrush — APPROVED** [review]

**Comment by cyrush** [comment]
Looks good!

**Pull request merged**